lib/Sema/SemaBounds.cpp

```diff
@@ -942,10 +942,15 @@ namespace {
       if (CK == CK_DynamicPtrBounds) {
         Expr *subExpr = E->getSubExpr();
         BoundsExpr *subExprBounds = S.InferRValueBounds(subExpr);
+        BoundsExpr *SrcBounds = S.InferRValueBounds(E);
 
         if (subExprBounds->isNone()) {
           S.Diag(subExpr->getLocStart(), diag::err_expected_bounds);
         }
+
+        assert(SrcBounds);
+        E->setBoundsExpr(SrcBounds);
```

Comment:
We should not overwrite the bounds expression that was declared by the user.   I have suggested another slot in CastExpr to hold these modified bounds.